### PR TITLE
updated python scripts in ci workflows with python files

### LINKED ISSUE #1148: Extract Python scripts out of GitHub workflows
## Problem

Because of their ease of use, the use of Python-based scripts inside GitHub workflows has increased a lot.

https://github.com/WordPress/openverse/blob/46a19a5776e747fd2b6111787222f2e6efc5afe3/.github/workflows/ci_cd.yml#L78

https://github.com/WordPress/openverse/blob/46a19a5776e747fd2b6111787222f2e6efc5afe3/.github/workflows/ci_cd.yml#L828

https://github.com/WordPress/openverse/blob/46a19a5776e747fd2b6111787222f2e6efc5afe3/.github/workflows/deploy-production-api.yml#L62

https://github.com/WordPress/openverse/blob/46a19a5776e747fd2b6111787222f2e6efc5afe3/.github/workflows/deploy-production-nuxt.yml#L62

https://github.com/WordPress/openverse/blob/46a19a5776e747fd2b6111787222f2e6efc5afe3/.github/workflows/deploy-staging-api.yml#L62

https://github.com/WordPress/openverse/blob/46a19a5776e747fd2b6111787222f2e6efc5afe3/.github/workflows/deploy-staging-nuxt.yml#L62

These Python scripts are not being linted, formatted or checked in any way. Contrast this with Bash scripts inside workflows that are linted by ShellCheck.

## Description

The solution is to keep them in a directory inside the codebase, possibly with `automations/python`.

## Additional context

- They can be kept separate from `automations/python` too, considering their usage, but these are all standalone scripts that do not need any packages to be installed.

- In fact, the last 4 Python scripts referenced above (in the ECS deployment workflows) are basically identical, and can be replaced by a single one that takes arguments using `argparse`.

### PR DESCRIPTION
**Pull request by rahulbollisetty**
## Fixes

Fixes #1148 by @dhruvkb 

## Description
- Updated Python scripts in the CI workflows with equivalent python files to improve formatting, linting.
- Added the python files in automations/python directory, which are mentioned in the workflows.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like
      `Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or
      a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible
      errors.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: .github/workflows/ci_cd.yml
Comment (dhruvkb) [inline .github/workflows/ci_cd.yml:78]: Since you're now invoking a Python script using `python <script.py>`, the shell must be changed to Bash (or dropped). This also applies to the other places with a similar change.

```suggestion
        shell: bash
```

or

```suggestion
```

**Review by dhruvkb — CHANGES_REQUESTED** [review]
This is great @rahulbollisetty and definitely on the right track. But there is some more work to be done:

- the Lint job has reported a lot of places where the Python code can be reformatted. Please run lint locally and implement the suggested changes.
- the shell must be changed to Bash (or dropped) in all the updated workflow jobs

File: .github/workflows/ci_cd.yml
Comment (rahulbollisetty) [reply .github/workflows/ci_cd.yml]: Sure, I have made the specified changes in the latest commit

Comment (rahulbollisetty) [reply .github/workflows/ci_cd.yml]: https://github.com/rahulbollisetty/openverse/blob/bc7bf3d40a7263a37230bde46356841a5bc3d85e/.github/workflows/ci_cd.yml#L73-L81

Comment (rahulbollisetty) [reply .github/workflows/ci_cd.yml]: https://github.com/rahulbollisetty/openverse/blob/bc7bf3d40a7263a37230bde46356841a5bc3d85e/.github/workflows/ci_cd.yml#L73-L81

https://github.com/rahulbollisetty/openverse/blob/bc7bf3d40a7263a37230bde46356841a5bc3d85e/.github/workflows/ci_cd.yml#L795-L805

https://github.com/rahulbollisetty/openverse/blob/bc7bf3d40a7263a37230bde46356841a5bc3d85e/.github/workflows/deploy-production-api.yml#L61-L64

https://github.com/rahulbollisetty/openverse/blob/bc7bf3d40a7263a37230bde46356841a5bc3d85e/.github/workflows/deploy-production-nuxt.yml#L61-L64

https://github.com/rahulbollisetty/openverse/blob/bc7bf3d40a7263a37230bde46356841a5bc3d85e/.github/workflows/deploy-staging-api.yml#L61-L64

https://github.com/rahulbollisetty/openverse/blob/bc7bf3d40a7263a37230bde46356841a5bc3d85e/.github/workflows/deploy-staging-nuxt.yml#L61-L64

Comment (rahulbollisetty) [reply .github/workflows/ci_cd.yml]: @dhruvkb I have made the suggested changes. Please have a look, and let me know if there are any more issues.

**Comment by rahulbollisetty** [comment]
@dhruvkb Why is there a failure in this check : 
https://github.com/WordPress/openverse/actions/runs/4636791825/jobs/8205030755?pr=1152
How to resolve it?

**Comment by dhruvkb** [comment]
@rahulbollisetty the action to add the stack label does not work on PRs from forks. It's a known issue #1090 and a fix is in the works. It won't be prevent this PR from being merged if that's your concern.

File: .github/workflows/ci_cd.yml
Comment (dhruvkb) [inline .github/workflows/ci_cd.yml:75]: Good catch, I had not thought of this.

File: .github/workflows/ci_cd.yml
Comment (dhruvkb) [inline .github/workflows/ci_cd.yml:81]: The `./` in the beginning is not necessary. It's not wrong, just not needed. Also applies to other similar lines.

```suggestion
        run: python automations/python/set_matrix_images.py
```

File: automations/python/generate_report.py
Comment (dhruvkb) [inline automations/python/generate_report.py:32]: Prefer to use parenthesis for longer strings over `\`. Somethign like this would be more readable.

```suggestion
                "text": (
                    f"<{server_url}/{repository}/actions/runs/{run_id}|"
                    "Click here to review the completed CI + CD workflow>."
                ),
```

File: automations/python/set_matrix_images.py
Comment (dhruvkb) [inline automations/python/set_matrix_images.py:1]:
```suggestion
"""Determine which Docker images to build and publish."""
```

File: automations/python/set_slack_user.py
Comment (dhruvkb) [inline automations/python/set_slack_user.py:1]:
```suggestion
"""Get Slack username corresponding to GitHub actor."""
```

File: automations/python/generate_report.py
Comment (dhruvkb) [inline automations/python/generate_report.py:1]:
```suggestion
"""Generate a completion report for the CI + CD workflow."""
```

**Review by dhruvkb — CHANGES_REQUESTED** [review]
This is looking really good and I like it a lot. There are only a few changes that I think can improve it further.

I think these Python files should be put inside a subfolder like `automation/python/workflows/`.

**Comment by rahulbollisetty** [comment]
> This is looking really good and I like it a lot. There are only a few changes that I think can improve it further.
> 
> I think these Python files should be put inside a subfolder like `automation/python/workflows/`.

I have updated all the changes , please look into it.

**Review by dhruvkb — APPROVED** [review]
Everything is looking really good, this is incredible. Thank you @rahulbollisetty! While we wait for another approval to merge, feel free to work on another issue.

**Comment by rahulbollisetty** [comment]
> Everything is looking really good, this is incredible. Thank you @rahulbollisetty! While we wait for another approval to merge, feel free to work on another issue.

Glad you liked the work @dhruvkb , Thank you so much for assisting me with the errors, would love to work on other django based issues.

**Comment by dhruvkb** [comment]
You should filter by "🧱 stack:api" or "🧱 stack:backend" label for those. Let us know if you need help finding an appropriate issue.

File: .github/workflows/deploy-production-api.yml
Comment (AetherUnbound) [inline .github/workflows/deploy-production-api.yml:1]: Per the header of this file, the `.github/workflows/deploy-*` workflows are all auto-generated from a different repository 😕 Unfortunately I think it would be best to keep they python code in-line here, unless we wanted to bring these changes upstream to the infrastructure repository which generates these files. What do you think @dhruvkb?

File: automations/python/workflows/generate_report.py
Comment (AetherUnbound) [inline automations/python/workflows/generate_report.py:16]: This can be replaced with the `int` type:
```suggestion
counts = defaultdict(int)
```

**Review by AetherUnbound — CHANGES_REQUESTED** [review]
Thanks for your work on this @rahulbollisetty! I think we might need to hold off on some of these changes, though, since they affect auto-generated files (noted below in a specific comment). Regardless of whether we decide to pull out the changes or keep things as is, we'd need to perform those changes in a separate PR. Do you mind removing `set_slack_user.py` and reverting the changes made to the `.github/workflows/deploy-*` files?

File: .github/workflows/deploy-production-api.yml
Comment (rahulbollisetty) [reply .github/workflows/deploy-production-api.yml:1]: So I have revert all the changes made and leave it as it was earlier,i.e, python code in-line in workflows?

Comment (AetherUnbound) [reply .github/workflows/deploy-production-api.yml:1]: Yes, but only for those workflows I mentioned. Your changes to `ci_cd.yml` are fantastic and should stay!

Comment (rahulbollisetty) [reply .github/workflows/deploy-production-api.yml:1]: sorry for the delay, I have updated the changes requested, please review it.

**Comment by AetherUnbound** [comment]
@rahulbollisetty we had some changes to the repo over the weekend, I've gone through and updated your PR based on those. Hope that's okay!

@dhruvkb I think I've captured the necessary changes as part of the rebase, do you mind taking a second look to make sure?

File: automations/python/workflows/set_slack_user.py
Comment (AetherUnbound) [inline automations/python/workflows/set_slack_user.py:1]: I _think_ we can remove this file too, since this was part of the deployment actions which are now reverted.

**Review by dhruvkb — APPROVED** [review]
This looks good to me! Thanks @rahulbollisetty for iterating on this so quickly and @AetherUnbound for rebasing and updating the PR.

**Comment by rahulbollisetty** [comment]
I have removed set_slack_user.py, and I see there are few commits containing bump-up for certain packages which are not in my repo (i.e., it's 5 commits behind), should I update them? @dhruvkb @AetherUnbound

**Comment by dhruvkb** [comment]
@rahulbollisetty updating is not needed. PRs can be merged even if they don't have some of the commits that have been added to `main` as long as there are no conflicting changes.

**Comment by rahulbollisetty** [comment]
> This looks good to me! Thanks @rahulbollisetty for iterating on this so quickly and @AetherUnbound for rebasing and updating the PR.

Happy to help!!

**Review by AetherUnbound — APPROVED** [review]
LGTM, thanks for your contribution!

**Comment by rahulbollisetty** [comment]
> LGTM, thanks for your contribution!

Glad you liked it, I'm happy to help,

**Pull request merged**
